### PR TITLE
Allow setting `history.state` when navigating

### LINKED ISSUE #10248: Router: Support Setting State Object When Navigating
In addition to setting query params, we should be able to set the state object. See here:

https://developer.mozilla.org/en-US/docs/Web/API/History_API

### LINKED ISSUE #25658: Allow to pass data with router.navigate
## I'm submitting a...
```
[x] Feature request
```

## Current behavior

Currently Angular does not provide a way of passing additional data with `router.navigate`.

There are many use cases for this, described in related issues which were unjustly closed without fulfilment:

- https://github.com/angular/angular/issues/5217
- https://github.com/angular/angular/issues/8515
- https://github.com/angular/angular/issues/9804
- https://github.com/angular/angular/issues/10248

#10248 is not actually closed, but addresses only a part of the problem.

## Expected behavior

Here is an example how it could work taken from https://github.com/angular/angular/issues/5217:

```
// exactly like @RouteConfig
this.router.navigateByUrl('/path', { data: { entity: 'entity' } })
```

The given `data` should be accessible everywhere where `route.data` is accessible (in the target component, resolvers if any etc.)

## Environment

Angular version: from 2 up to the current.

## Additional Information

There is a workaround by @Hipparch:
https://github.com/Hipparch/Angular2-navigate-with-data

### PR DESCRIPTION
**Pull request by jasonaden**
This feature adds a few capabilities. First, when a `popstate` event fires the value of `history.state` will be read and passed into `NavigationStart`. In the past, only the `navigationId` would be passed here.

Additionally, `NavigationExtras` has a new public API called `state` which is any object that will be stored as a value in `history.state` on navigation. For example, the object `{foo: 'bar'}` will be written to `history.state` here:

`router.navigateByUrl('/simple', {state: {foo: 'bar'}});`

Additionally, the `NavigationExtras` can now be read during navigation. We're exposing a new type `Navigation` that provides information such as the navigation ID, target URL, `NavigationExtras` passed to the `navigateByUrl` method, and a string value indicating what triggered the navigation (popstate, imperative, etc).

In order to read this `Navigation` object, use the `router.getCurrentNavigation()` function. If there's no navigation currently executing, the value will be `null`. Continuing from the above example:

```
navigation = router.getCurrentNavigation();

expect(navigation.trigger).toBe('imperative');
expect(navigation.extras.state).toEqual({foo: 'bar'});`
```

These features are related to a series of issues regarding adding arbitrary data to navigations (#5217 #8515 #9804 #24617). Additionally, this PR closes #10248 for supporting setting `history.state`.

closes #25658

### CODE REVIEW

File: packages/router/src/router.ts [hunk 1]
```diff
@@ -917,7 +936,7 @@ export class Router {
 
   private scheduleNavigation(
       rawUrl: UrlTree, source: NavigationTrigger, restoredState: {navigationId: number}|null,
-      extras: NavigationExtras): Promise<boolean> {
+      futureState: {[key: string]: any}|null, extras: NavigationExtras): Promise<boolean> {
```
Comment (jasonaden) [inline packages/router/src/router.ts]: Remove `futureState`

File: packages/router/src/router.ts [hunk 2]
```diff
@@ -966,12 +985,14 @@ export class Router {
...
   }
 
-  private setBrowserUrl(url: UrlTree, replaceUrl: boolean, id: number) {
+  private setBrowserUrl(
+      url: UrlTree, replaceUrl: boolean, id: number, state?: {[key: string]: any}) {
     const path = this.urlSerializer.serialize(url);
+    state = state || {};
     if (this.location.isCurrentPathEqualTo(path) || replaceUrl) {
-      this.location.replaceState(path, '', {navigationId: id});
+      this.location.replaceState(path, '', {...state, navigationId: id});
```
Comment (jasonaden) [inline packages/router/src/router.ts]: Update with `ng` namespace.

File: packages/router/src/router.ts [hunk 3]
```diff
@@ -253,6 +301,7 @@ export class Router {
 
   public readonly events: Observable<Event> = new Subject<Event>();
   public readonly routerState: RouterState;
+  public currentTransition: Transition|null = null;
```
Comment (jasonaden) [inline packages/router/src/router.ts]: `private`

File: packages/router/src/router.ts [hunk 4]
```diff
@@ -181,6 +181,53 @@ function defaultMalformedUriErrorHandler(
...
+export type Transition = {
+  /**
+   * The ID of the current navigation.
+   */
+  id: number;
+  /**
+   * Raw target URL passed by the user when executing navigation. This is the value before the
+   * router has parsed or applied redirects to it.
+   */
+  rawUrl: string | UrlTree;
```
Comment (jasonaden) [inline packages/router/src/router.ts]: `initialUrl`

File: packages/router/src/router.ts [hunk 5]
```diff
@@ -181,6 +181,53 @@ function defaultMalformedUriErrorHandler(
...
+  /**
+   * The raw target URL after being parsed with UrlSerializer.extract(). This is the most
+   * commonly used URL when determining if two URLs match.
+   */
+  extractedUrl: UrlTree;
+  /**
+   * Target URL after redirects have been applied. This URL may not be available immediately,
+   * therefore this property can be `undefined`.
+   */
+  url?: UrlTree;
```
Comment (jasonaden) [inline packages/router/src/router.ts]: `finalUrl`

**Comment by jasonaden** [comment]
[Presubmit](http://test/OCL:223393780:BASE:223393849:1543524410774:b62e2bb3)

**Comment by alxhub** [comment]
LGTM

The only thing I would add is that the commits here could benefit from additional info (e.g. a description of what the old behavior was and the motivation for changing it). One useful thing to add in each commit message in a multi-commit PR like this is a line or two about how the code in the commit is tested.

**Review by alxhub — APPROVED** [review]

File: packages/router/src/router.ts [hunk 6]
```diff
@@ -169,6 +181,57 @@ function defaultMalformedUriErrorHandler(
...
+   */
+  finalUrl?: UrlTree;
+  /**
+   * Identifies the trigger of the navigation.
+   *
+   * * 'imperative'--triggered by `router.navigateByUrl` or `router.navigate`.
+   * * 'popstate'--triggered by a popstate event
+   * * 'hashchange'--triggered by a hashchange event
+   */
+  source: 'imperative' | 'popstate' | 'hashchange';
```
Comment (IgorMinar) [inline packages/router/src/router.ts]: source could be understood as the source url of the transition. In the description you described it as a "trigger" that tells me that "trigger" could be a better name?

File: packages/router/src/router.ts [hunk 7]
```diff
@@ -169,6 +181,57 @@ function defaultMalformedUriErrorHandler(
...
+  /**
+   * The NavigationExtras used in this navigation. See {@link NavigationExtras} for more info.
+   */
+  extras: NavigationExtras;
+  /**
+   * Previously successful Transition object. Only a single previous transition is available,
+   * therefore this previous Transition will always have a `null` value for `previousTransition`.
+   */
+  previousTransition: Transition | null;
```
Comment (IgorMinar) [inline packages/router/src/router.ts]: what's the purpose of exposing this? if it's just to provide access to the previous url, then why not expose that directly as `previousUrl`?

File: tools/public_api_guard/router/router.d.ts [hunk 8]
```diff
@@ -185,19 +185,29 @@ export interface NavigationExtras {
...
     skipLocationChange?: boolean;
+    state?: {
+        [k: string]: any;
+    };
 }
 
 export declare class NavigationStart extends RouterEvent {
     navigationTrigger?: 'imperative' | 'popstate' | 'hashchange';
-    restoredState?: {
+    /** @deprecated */ restoredState?: {
```
Comment (IgorMinar) [inline tools/public_api_guard/router/router.d.ts]: see.. this is wrong. :-|

File: tools/public_api_guard/router/router.d.ts [hunk 9]
```diff
@@ -185,19 +185,29 @@ export interface NavigationExtras {
...
+        [k: string]: any;
         navigationId: number;
+        ng: {
```
Comment (IgorMinar) [inline tools/public_api_guard/router/router.d.ts]: why did you decide to create a separate namespace for this? because you are afraid of collisions with user defined properties? do we have a need to store more in here besides the id? (any future plans?)

File: tools/public_api_guard/router/router.d.ts [hunk 10]
```diff
@@ -464,6 +481,16 @@ export declare class Scroll {
     toString(): string;
 }
 
+export declare type Transition = {
+    id: number;
```
Comment (IgorMinar) [inline tools/public_api_guard/router/router.d.ts]: isn't it odd that we have `navigationId` all over the place and Transition#id is set to `navigationId`? seems like an api consistency issue..

**Review by IgorMinar — CHANGES_REQUESTED** [review]

File: packages/router/src/directives/router_link.ts [hunk 11]
```diff
@@ -77,6 +77,27 @@ import {UrlTree} from '../url_tree';
...
+ * ```
+ *
+ * And later the value can be read from the router through `router.getCurrentTransition.
+ * For example, to capture the `tracingId` above during the `NavigationStart` event:
+ *
+ * ```
+ * // Get NavigationStart events
+ * router.events.pipe(filter(e => e instanceof NavigationStart)).subscribe(e => {
+ *   const transition = router.getCurrentTransition();
+ *   tracingService.trace({id: transition.extras.state});
```
Comment (jasonaden) [inline packages/router/src/directives/router_link.ts]: ...`state.tracingId`

File: packages/router/src/router.ts [hunk 12]
```diff
@@ -169,6 +181,57 @@ function defaultMalformedUriErrorHandler(
...
+   */
+  id: number;
+  /**
+   * Target URL passed by the user when executing navigation. This is the value before the
+   * router has parsed or applied redirects to it.
+   */
+  initialUrl: string | UrlTree;
+  /**
+   * The initial target URL after being parsed with UrlSerializer.extract(). This is the most
+   * commonly used URL when determining if two URLs match.
```
[the remark this answers is not in the document]
Comment (jasonaden) [reply packages/router/src/router.ts]: Removed. This is possibly a long explanation, and if they are using any special extraction, they will understand anyway. I had earlier started documenting it and it was way too wordy and confusing for this part of the docs.

File: packages/router/src/router.ts (hunk 7 above)
Comment (jasonaden) [reply packages/router/src/router.ts]: Because then we have to have 3 copies. The `previousInitialUrl`, `previousExtractedUrl` and `previousFinalUrl`. Also, we may add more data to this over time and exposing it this way means we don't need to bloat the object.

**Review by IgorMinar — APPROVED** [review]

**Comment by StevenLiekens** [comment]
Thanks, I've been waiting for this. Appreciate the effort.

**Comment by trotyl** [comment]
@jasonaden Should scroll positions also be stored on history state based on this feature? Current object map is technically a memory leak that grows with navigations.

**Comment by Virendra-Yadav** [comment]
https://www.npmjs.com/package/ngx-navigation-with-data
For more details

**Comment by StevenLiekens** [comment]
I don't pretend to be a moderator but that's like the fifth time you advertised your package instead of commenting on the issue. Stop it.

**Comment by jasonaden** [comment]
@trotyl It probably should be. I have an internal issue (just created [this issue](https://github.com/angular/angular/issues/27607) to track it) to deprecate `navigationId` in `history.state`. Then we will have a key we can store Angular-specific data under. At that point, yes, we can get scroll positions in there. This would have the advantage of also persisting scroll position through navigations.

If anyone comes across this and wants to work on the issue, that would be great. I should be getting to it within a couple weeks if not.

**Comment by Airblader** [comment]
@jasonaden This feature looks really cool, thanks. However, as a user I am slightly confused about the lifespan of a "current navigation". If I subscribe to the queryParamMap, is it safe to rely on the current navigation's state as long as I stay synchronous (i.e., no switchMap's to some async observable)? I can't seem to find any navigation regarding the lifecycle of a single navigation.

**Comment by sky10p** [comment]
It does not seem to work, when I use getCurrentNavigation it always returns null.

I leave the test that I have done:

https://stackblitz.com/edit/angular-wgc6ou

Thanks!

**Comment by fjauer** [comment]
@sky10p i got your example working:

just moved the coding from ngOnInit to the constructor

https://stackblitz.com/edit/angular-bupuzn

**Comment by sky10p** [comment]
@fjauer And why does that work? Does the constructor not run before?

Thanks for your time!

**Comment by alecvolo** [comment]
I think there is a small bug - the `history.state` object is getting lost when navigate using back/forward buttons.
The fix should be pretty simple- set restored state into the `NavigationExtras `  parameter in the call of `scheduleNavigation` in `Router.setUpLocationChangeListener()` method :


```javascript
        // Navigations coming from Angular router have a navigationId state property. When this
        // exists, restore the state.
        const state = change.state && change.state.navigationId ? change.state : null;
     setTimeout(
            () => { this.scheduleNavigation(rawUrlTree, source, state, {replaceUrl: true, state }); }, 0);
```

The workaround is to catch `NavigationStart` event and fire navigation from there:
```javascript
    this._router.events.pipe(
        filter(t =>t instanceof NavigationStart && (<NavigationStart>t).navigationTrigger == "popstate" ))
      .subscribe((x: NavigationStart) => {
        this._router.navigate([x.url], { state: x.restoredState });
      });
```

**Comment by Airblader** [comment]
@alecvolo Please open a (new) issue instead of commenting on closed ones.

**Comment by f-aubert** [comment]
Maybe a stupid question? But was this nice feat merged in any official Angular Release? 7.2? Or does it wait for Angular 8?

**Comment by f-aubert** [comment]
Thanks @Airblader for pointing me to the ChangeLog. I did miss it. I would have expected  this Issue to be somehow tagged with 7.2.0, and the Routing Documentation to reflect this feature. But many thanks to all that contributed, it was greatly awaited and needed in my apps.

**Comment by nhays89** [comment]
Probably should create a new bug, but just wanted to see if this should work when receiving an error response in an interceptor and trying to pass the error message to an error component via the router's state object? I am getting a null object as @sky10p described from the call to getCurrentNavigation() and the code is in the constructor as @fjauer commented.

**Comment by jasonaden** [comment]
Thanks for the feedback on this issue everyone. Good that it's useful for many, sorry about any problems with documentation or problems using the feature.

`getCurrentNavigation()` will return the *currently executing* navigation. It will return *null* if there is no currently executing navigation (in other words, the application is in a stable state).

Internally, there is a reference to `lastSuccessfulNavigation`, but this isn't exposed publicly. We tend to err on the side of not publicly exposing APIs unless valid use cases are presented. It sounds a bit like what some of you are looking for is an API to get either the currently executing navigation or the previous one.

If someone wants to create a feature request for this with use cases, that would help clarify.

Thanks!

**Comment by cogidoo** [comment]
Why is this only implemented for the `RouterLinkWithHref`-directive. So currently it's not possible to use the state-feature with a routerlink-directive at - let's say - a button.

**Comment by colmben** [comment]
By saying this works only on the currently executing navigation, does that mean it shouldn't be used outside of the navigation lifecycle hooks?  It doesn't work by the time ngInit is called in a component for example, getCurrentNavigation has reverted to null by that point. The stackblitz above shows it does work in a component constructor, but is that guaranteed? If it is used inside a service subscription in the constructor for example, will it have reverted to null by the time it is called? 

In general, is this approach supposed to replace the paramsMap.get method of passing params? I see it linked to a lot of those issues but the actual code here seems more related to the nav lifecycle hooks than giving the component access to the state?

**Comment by Airblader** [comment]
> In general, is this approach supposed to replace the paramsMap.get method of passing params?

I'm not from the Angular team, but I'd confidently say »No«. This additional state has nothing to do with route or query parameters, and accessing those is always possible from an ActivatedRoute (even outside a navigation).

> The stackblitz above shows it does work in a component constructor, but is that guaranteed?

This is definitely an interesting question. I can't find any kindd of documentation or explanation as to what the design of on-going navigations and their lifecycle is. It's just a "it's on-going until it finished", which isn't very satisfying.

What I can tell you is where I already use this feature: I subscribe to query parameters and inside that subscription I access the current navigation to check the state for a specific flag.

**Comment by colmben** [comment]
> > In general, is this approach supposed to replace the paramsMap.get method of passing params?
> 
> I'm not from the Angular team, but I'd confidently say »No«. This additional state has nothing to do with route or query parameters, and accessing those is always possible from an ActivatedRoute (even outside a navigation).

By paramsMap.get I more meant the optional params that are passed via the matrix method in routing. This new functionality here is already being touted on StackOverflow as the new method for doing data passing via routing:

https://stackoverflow.com/a/54365098/1669024

I can see why people would jump on it for optional params as it means they aren't getting tacked onto the URL (and are available to all segments of the route, not just the last as with the matrix method).

**Comment by Airblader** [comment]
> I can see why people would jump on it for optional params as it means they aren't getting tacked onto the URL

Yes, for that state would be a solution. I'd question whether you *want* to hide it from the URL, though, since encoding it on the URL makes the link reentrant (but [I'm biased](https://ngqp.io)).

**Comment by colmben** [comment]
> Yes, for that state would be a solution. I'd question whether you _want_ to hide it from the URL, though, since encoding it on the URL makes the link reentrant (but [I'm biased](https://ngqp.io)).

And indeed that comes up in the SO comments :) In general though I think people are not liking those massive URLs...

**Comment by sagarjaspal** [comment]
> 
> 
> Why is this only implemented for the `RouterLinkWithHref`-directive. So currently it's not possible to use the state-feature with a routerlink-directive at - let's say - a button.

I tried it with mat-card, didn't work :(

**Comment by eric-gagnon** [comment]
Hello,

It's probably more of a support question but I'm really lost in documentation, tutorials, fixed or partially fixed bugs. Also very new to Angular.

I want to implement a sane handling of back / forward / reload using url parameters but I'm facing a cas where I want to keep track of a state but not expose it in the url (in this case the return url to original search form page). Back cannot be used and using service to persist the state has quite a few corner case. I tought that using the state from history api would let me implement this but it's not working like I expect.

In Angular 7.2.x I'm able to pass the state and read it (window.history.state) but it's not recoved when reloading the page. Is there something I'm missing or it's been fixed in more recent version?

Ex. From first route

this._router.navigateByUrl('/somewhere', { state: { somestate: "test" } });

From second route  / comonent

this.state$ = this.activatedRoute.paramMap.pipe(map(() => window.history.state))

    this.state$.subscribe((val: any) => {
      console.log(val); // Ok I see the state.
    });

On browser reload, console.log only display a state with navigation id property.

Maybe I need to do something to force the persisting in history or the value is somewhere else when angular is restarting from reload (hard to find using chrome as most the values are now observables)? Or maybe I'm trying to do something that is simply not possible in Angular?

Would take any pointer to relevant thread or hint on how to do it properly!

**Comment by SzymonKatra** [comment]
> > Why is this only implemented for the `RouterLinkWithHref`-directive. So currently it's not possible to use the state-feature with a routerlink-directive at - let's say - a button.
> 
> I tried it with mat-card, didn't work :(

Is it going to be fixed? I use `RouterLink` on `button` and state-feature doesn't work either.

**Comment by SzymonKatra** [comment]
> > Why is this only implemented for the `RouterLinkWithHref`-directive. So currently it's not possible to use the state-feature with a routerlink-directive at - let's say - a button.
> 
> I tried it with mat-card, didn't work :(

Is it going to be fixed? I wanted to use `routerLink` with state-feature on `button` but with no success.
